src/MongoDB.Driver/Encryption/ClientEncryption.cs

```diff
@@ -0,0 +1,194 @@
+        /// </summary>
+        /// <param name="value">The value.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The decrypted value.</returns>
+        public async Task<BsonBinaryData> DecryptAsync(BsonBinaryData value, CancellationToken cancellationToken)
+        {
+            var wrappedValueBytes = GetWrappedValueBytes(value);
+            return await _libMongoCryptController.DecryptFieldAsync(wrappedValueBytes, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Encrypts the specified value.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        /// <param name="encryptOptions">The encrypt options.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The encrypted value.</returns>
+        public BsonBinaryData Encrypt(BsonValue value, EncryptOptions encryptOptions, CancellationToken cancellationToken)
+        {
+            var unencryptedWrappedValueBytes = GetWrappedValueBytes(value);
+            GetEncryptFieldArgs(encryptOptions, out var keyId, out var alternateKeyNameBytes, out var algorithm);
+            var encryptedWrappedValueBytes = _libMongoCryptController.EncryptField(
+                unencryptedWrappedValueBytes,
+                keyId,
+                alternateKeyNameBytes,
+                algorithm,
+                cancellationToken);
+            return UnwrapEncryptedValue(encryptedWrappedValueBytes);
+        }
+
+        /// <summary>
+        /// Encrypts the specified value.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        /// <param name="encryptOptions">The encrypt options.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The encrypted value.</returns>
+        public async Task<BsonBinaryData> EncryptAsync(BsonValue value, EncryptOptions encryptOptions, CancellationToken cancellationToken)
+        {
+            var unencryptedWrappedValueBytes = GetWrappedValueBytes(value);
+            GetEncryptFieldArgs(encryptOptions, out var keyId, out var alternateKeyNameBytes, out var algorithm);
+            var encryptedWrappedValueBytes = await _libMongoCryptController
+                .EncryptFieldAsync(
+                    unencryptedWrappedValueBytes,
+                    keyId,
+                    alternateKeyNameBytes,
+                    algorithm,
+                    cancellationToken)
+                .ConfigureAwait(false);
+            return UnwrapEncryptedValue(encryptedWrappedValueBytes);
+        }
+
+        private void GetEncryptFieldArgs(
+            EncryptOptions encryptOptions,
+            out Guid? keyId,
+            out byte[] alternateKeyNameDocuments,
+            out EncryptionAlgorithm algorithm)
+        {
+            keyId = encryptOptions.KeyId != null ? new Guid(encryptOptions.KeyId) : (Guid?)null;
```

Comment:
Then there must be an equal and opposite bug somewhere else where the keys are created.

I suspect this line:

https://github.com/DmitryLukyanov/mongo-csharp-driver/blob/csharp2431/src/MongoDB.Driver/Encryption/LibMongoCryptController.cs#L118

which is very likely corrupting the representation of the `_id` when inserting it into the collection.

You should write a test that asserts that the `_id` in the collection has the correct representation (and value).

Reply:
Do you have any reason to suspect that `encryptOptions.keyId` is NOT in `Standard` byte order?

Reply:
I will come back to this question a little bit later today

Reply:
I am 99.9% certain you will need to change this to:

    GuidConverter.FromBytes(encryptOptions.KeyId, GuidRepresentation.Standard)

If this causes tests to fail it will be because of a compensating bug somewhere else.